client/dfclient/src/gameplay_state_death_reports.cpp

```diff
@@ -0,0 +1,449 @@
+/** 
+ * The processDeathReport() procedure and everything that goes with it.
+ * This includes handling the killcount mechanics. Killcount mechanics are:
+ *  - killing spree - multiple kills without dying
+ *  - shutdown - stopping somebody's killing spree
+ *  - domination - killing a player without him getting a kill back
+ *  - revenge - getting a kill back at a dominating player
+ *  - comeback - getting a kill for the first time after dying multiple times
+ *  - multikill - getting several kills in a quick succession
+ */
+
+#include <ncine/Application.h>
+
+#include "../../../common/constants.hpp"
+
+#include "game_data.hpp"
+#include "gameplay_state.hpp"
+#include "resources.hpp"
+#include "text_creator.hpp"
+
+// translation-unit-local variables used throughout the code in this file
+static std::optional<TextCreator> theTextCreator;	// instanced here for potential future thread-safety
+static ncine::SceneNode& rootNode = ncine::theApplication().rootNode();
+static float screenWidth;
+static float screenHeight;
+static std::array<float, 3> textsXPos;
+static std::array<float, 4> textsYPos;
```

Comment:
No, please, let's not add globals if it's not necessary.

If you need to keep some state across several function calls, why not wrap those globals into a "context" structure, or create a new class for the sole purpose of processing the death report?

Reply:
As far as I can tell, these variables are global and visible only in the scope of this translation unit (for example, I cannot use them in gameplay_state.cpp, which is in the same place in the file dependency hierarchy). Is this really so bad? These are just helpers, universal variables that I use in every function in this file. Making them global allows me to use them in these functions without having to declare them each time. If there is some more evil coming from them being global, please tell me, so I can learn :D But I can't see anything wrong here.

Ideally they would be `const`, but they cannot be initialized at the start of the program, because their values (screen width, for example) are not specified yet. If you have a better idea how to make them visible in all the functions (and `const`), please tell. :D

Reply:
> As far as I can tell, these variables are global and visible only in the scope of this translation unit (for example, I cannot use them in gameplay_state.cpp, which is in the same place in the file dependency hierarchy).

OK, you are right that the variables are not global - their scope is restricted to only one file. In such, case, I have two problems with them:

- They make the functions in this file not thread-safe. Not sure if a problem in this case, but it is in general. We can't really say that those variables will always be initialized to the same value, because the screen (window) size can change.
- This
  ```
  static ncine::SceneNode& rootNode = ncine::theApplication().rootNode();
  ```
  is dangerous, because it's prone to the [static initialization order fiasco](https://en.cppreference.com/w/cpp/language/siof). If you have statically-initialized variables in file A which depend in some way (either directly or indirectly, e.g. through a function) on static variables in file B, then there is a risk that variables in A will be attempted to be initialized before those in file B.

What I suggest to do is to create a class/struct which will live for the duration of processing the death report. The helper static variables can be converted into `const` members, and initialized in the constructor. Something like this:

```c++
// Somewhere in the beginning of the gameplay_state_death_reports.cpp file:
class DeathReportProcessor {
private:
    const TextCreator _textCreator;
    const ncine::SceneNode& _rootNode;
    const float _screenWidth, _screenHeight;
    const std::array<float, 3> _textsXPos;
    const std::array<float, 4> _textsYPos;

public:
    DeathReportProcessor(Resources& resources)
        : _textCreator(resources)
        , _rootNode(ncine::theApplication().rootNode())
        , _screenWidth(ncine::theApplication().width())
        , _screenHeight(ncine::theApplication().height())
        , _textsXPos{0.41f * _screenWidth, 0.5f * _screenWidth, 0.59f * _screenWidth}
        , _textsYPos{0.75f * _screenHeight, 0.5f * _screenHeight, 0.30f * _screenHeight, 0.25f * _screenHeight}
    {}

    // Main method declaration
    void ProcessDeathReport(FlatBuffGenerated::DeathReport* report);

private:
    // Convert other helper methods and declare them here
    void processKill(const FlatBuffGenerated::DeathReport* deathReport);
    std::unique_ptr<ncine::TextNode> processMultikill(int multikillness);
    void endKillingSpree();
    // et cetera
};

void GameplayState::ProcessDeathReport(void const* ev) {
    const auto deathReport = (const FlatBuffGenerated::DeathReport*)ev;
    DeathReportProcessor processor(_resources);
    processor.ProcessDeathReport(ev);
}

// Here go the definitions of DeathReportProcessor's methods...
```

This way you can still access the variables from the helper functions without having to prefix them in any way (only with underscore), and now you have a separate instance of those variables for each ProcessDeathReport, so there is no risk of accidental sharing, and no risk of static order initialization fiasco.

Reply:
Well, that seems to be our golden answer to all the problems. Thanks for all the help, will implement :)